1_fetch/src/get_nwis_data.R

```diff
@@ -1,16 +1,35 @@
 
-download_nwis_data <- function(site_nums = c("01427207", "01432160", "01435000", "01436690", "01466500", "01477050"), parameterCd = '00010'){
-  
-  # create the file names that are needed for download_nwis_site_data
+download_nwis_site_data <- function(site_nums = c("01427207", "01432160", "01435000", "01436690", "01466500", "01477050"), parameterCd = '00010', startDate="2014-05-01", endDate="2015-05-01"){
   # tempdir() creates a temporary directory that is wiped out when you start a new R session; 
   # replace tempdir() with "1_fetch/out" or another desired folder if you want to retain the download
-  download_files <- file.path(tempdir(), paste0('nwis_', site_nums, '_data.csv'))
-  
+  download_files <- file.path("1_fetch/out", paste0('nwis_', site_nums, '_data.csv'))
+
+  for (filepath in download_files) {
+    # filepaths look something like directory/nwis_01432160_data.csv,
+    # remove the directory with basename() and extract the 01432160 with the regular expression match
+    site_num <- basename(filepath) %>% 
+      stringr::str_extract(pattern = "(?:[0-9]+)")
+    
+    # readNWISdata is from the dataRetrieval package
+    data_out <- readNWISdata(sites=site_num, service="iv", 
+                             parameterCd = parameterCd, startDate = startDate, endDate = endDate)
+    
+    # -- simulating a failure-prone web-sevice here, do not edit --
+    if (sample(c(T,F,F,F), 1)){
+      stop(site_num, ' has failed due to connection timeout. Try scmake() again')
+    }
+    # -- end of do-not-edit block
+    
+    write_csv(data_out, path = filepath)
+  }
+  return(download_files)
```

Comment:
download files is a vector of file names. This will be the right inputs for the following function, but will it serve the purpose of triggering a rebuild of `site_data` if the contents of the files changes? 

This is a great start, but because of the makeup of the `raw_downloads` target, it will only look out of date to scipiper if the file names change, the number of files changes, or the order of the files changes
```
"1_fetch/out/nwis_01427207_data.csv" "1_fetch/out/nwis_01432160_data.csv" "1_fetch/out/nwis_01435000_data.csv" "1_fetch/out/nwis_01436690_data.csv" "1_fetch/out/nwis_01466500_data.csv" "1_fetch/out/nwis_01477050_data.csv"
```
But, in the case where one data value in one of these files has changed but nothing else has changed, we won't see the approach trigger a rebuild

Did you consider creating a target for each file, like:
```
1_fetch/out/nwis_01427207_data.csv:
   command: download_nwis_site_data(target_name)
1_fetch/out/nwis_01432160_data.csv:
   command: download_nwis_site_data(target_name)
```
I know that is a painful amount of typing and targets, but it might offer a couple advantages: 1) any changes to the data are appropriately tracked by scipiper, 2) the failures are isolated to each file, so it is easier to build all of them progressively than luck out and be able to build all files at once.

Reply:
Yes I can do that. I'd thought of that approach, but didn't know if it would create too many targets. Also - wouldn't that approach be unwieldy if you had a bunch of sites for which you were downloading data?

Reply:
Yes, the approach isn't scalable, but it works for this case. Cool/fun solutions for scaling in pipelines III 🎉

Reply:
>the failures are isolated to each file, so it is easier to build all of them progressively than luck out and be able to build all files at once

That seems like a big pro to this approach.